README.txt
Description
===========

*Please send me feedback and comments to carlitos.kyo@gmail.com*

Based on djbuild and code from setuptools used.

This buildout recipe can be used to create a setup for gae project. It will
automatically install apps into libs directory.

You can see an example of how to use the recipe below::

  [buildout]
  parts = satchmo gae
  eggs = ipython

  [satchmo]
  recipe = gocept.download
  url = http://www.satchmoproject.com/snapshots/satchmo-0.6.tar.gz
  md5sum = 659a4845c1c731be5cfe29bfcc5d14b1

  [gae]
  recipe = gaebuild
  settings = development
  eggs = ${buildout:eggs}
  project = dummyshop


Supported options
=================

The recipe supports the following options.

apps 
  projects that can be installed using pypi or compressed files. No handle
  dependencies do it by hand using buildout, the decision was taken for these reasons:
  
  * if dependency is a gae/django app this should be declared into this option to install
    it into the extarnal-apps directory or it should be omited if the dependency
    was customized and it is on local-apps directory
    
  * if dependency is not a gae/django app this should be declared into libs option.
  
  To delete an application should be by hand.

project
  This option sets the name for your project. The recipe will create a
  basic structure if the project is not already there.
  
external-apps
  This option sets the directory where external reusable apps goes. Which do not
  be installed as an egg or if you don't want install it as an egg.
  
local-apps
  This option sets the directory where local reusable apps goes, usually
  put the company name for this directory, and customized apps.

python
  This option can be used to specify a specific Python version which can be a
  different version from the one used to run the buildout.

settings
  You can set the name of the settings file which is to be used with
  this option. This is useful if you want to have a different
  production setup from your development setup. It defaults to
  `development`. if turboengine is setup

download-cache
  Set this to a folder somewhere on you system to speed up
  installation. The recipe will use this folder as a cache for a
  downloaded version of Django.

libs-path
  Path specified here will be used to install python projects
  
script-dir
  Paths specified here will be used to extend the default Python
  path for the `bin/*` scripts. default 'bin' some common scripts
  will be wsdl2py used for webservices and ZSI

find-links
  used to install apps inside the project
  
turboengine
  use turboengine set the version or 'last' for last version
  
webservices
  set true to use turboengines webservices wrapper, gaebuild will install ZSI 
  and zope.interface only avaible if turboengine is used, default false
  
libs
  pure python libraries to install into libs-dir, default lib
  
zipped
  true for install compressed eggs(libs) format or false as directory default false.

All following options only have effect when the project specified by
the project option has not been created already, on the setting file 
especified.

Notes.-

* If use turboengine don't edit app.py if generated authomaticaly by the recipe but Please
  edit webservices.py only generated once.
* If not edit app.py first extending pythonpath if you change this variables: local-apps, extarnal-apps
  libs-dir
* If some app or project has error on its setup.py file the process is terminated


Another example
===============

The next example shows you how to use some more of the options::

  [buildout]
  parts = gae extras
  eggs =
    hashlib

  [extras]
  recipe = iw.recipe.subversion
  urls =
    http://django-command-extensions.googlecode.com/svn/trunk/ django-command-extensions
    http://django-mptt.googlecode.com/svn/trunk/ django-mptt

  [gae]
  recipe = gaebuild
  settings = development
  project = exampleproject
  eggs =
    ${buildout:eggs}


Example with a different Python version
=======================================

To use a different Python version from the one that ran buildout in the
generated script use something like::

  [buildout]
  parts	= myproject

  [special-python]
  executable = /some/special/python

  [myproject]
  recipe	= gaebuild
  project	= myproject
  python	= special-python


CHANGES.txt
Changes
=======
25-Oct-2010: fixing bug find-link option now into the correct place to install
25-Oct-2010: adding base templates base.html, 404.html, 500.html
25-Oct-2010: fixing templates
28-Oct-2010: fixing templates for generating correct scripts
28-Oct-2010: adding local_settings.py code less in setting options
29-Oct-2010: fixing settings.py file to import correctly
27-jun-2011: updating to new turboengine code
07-Oct-2011: new base.html template HTML5


src/gaebuild/gaebuild.py
import logging, os, zc.buildout
import os

from installer import Installer

class GAEBuild(object):

    def __init__(self, buildout, name, options):
        self.log = logging.getLogger(name)
        self.egg = zc.recipe.egg.Egg(buildout, options['recipe'], options)

        self.buildout, self.name, self.options = buildout, name, options
        options['location'] = os.path.join(
            buildout['buildout']['parts-directory'], name)
        options['bin-directory'] = buildout['buildout']['bin-directory']

        options.setdefault('project', 'example.com')
        options.setdefault('external-apps', 'apps')
        options.setdefault('local-apps', 'company')
        options.setdefault('settings', 'development')
        options.setdefault('libs-path', 'lib')
        options.setdefault('script-dir', 'bin')
        options.setdefault('turboengine', '')
        options.setdefault('webservices', 'false')
        options.setdefault('zipped', 'false')

        # Usefull when using archived versions
        buildout['buildout'].setdefault(
            'download-cache',
            os.path.join(buildout['buildout']['directory'],
                         'downloads'))
        
        self.__installer = Installer(self.options, self.buildout, self.log, self.name)


    def install(self):
        print '\n------------------ Installing GAEBuild ------------------\n'
        location = self.options['location']
        base_dir = self.buildout['buildout']['directory']
        project_dir = os.path.join(base_dir, 'src')
        download_dir = self.buildout['buildout']['download-cache']
        
        extra_path = self.__installer.get_extra_paths()
        requirements, ws = self.egg.working_set(['gaebuild'])
        
        self.__installer.verify_or_create_download_dir(download_dir)
        
        self.__installer.install_recipe(location)
        self.__installer.install_project(project_dir, self.options['project'])
        
        script_paths = self.__installer.install_scripts(project_dir, extra_path, ws)
        
        print '\n------------------ Installing GAEBuild ------------------\n'

        return script_paths + [location]

    def update(self):
        pass
    

src/gaebuild/templates.py
production_settings = '''from settings import *

DEBUG = False

# url, callschuduler
INSTALLED_APPS += [
    #('/', 'app'),
    #('/other/', 'app'),
]
'''

development_settings = '''from settings import *

DEBUG = True

# url, callschuduler
INSTALLED_APPS += [
    #('/', 'app'),
    #('/other/', 'app'),
]
'''

configs = {
    'turboengine': {
        'app_py': '''#########################################################
# Extending python path, file generated not edit
#########################################################

def update_path():
    import os
    import sys

    base_dir = os.path.abspath(os.path.dirname(__file__))

    local_dir = os.path.join(base_dir,'%(local)s/')
    external_dir = os.path.join(base_dir,'%(external)s/')

    lib_dir = os.path.join(base_dir,'%(lib)s/')

    if lib_dir not in sys.path:
        import site
        site.addsitedir(lib_dir)


    if local_dir not in sys.path:
        sys.path.insert(0, local_dir)

    if external_dir not in sys.path:
        sys.path.insert(0, external_dir)

update_path()

#########################################################
# End extending python path, file generated not edit
#########################################################

from google.appengine.ext import webapp
from google.appengine.ext.webapp.util import run_wsgi_app

def main(application):
    run_wsgi_app(application)

if __name__ == "__main__":
    # first setting which file use
    import os
    os.environ['turboengine.settings'] = 'project.%(settings)s'

    # then we can use turboengine
    from turboengine import urls
    from turboengine.conf import settings

    application = webapp.WSGIApplication(urls.generate_urls(settings.INSTALLED_APPS), debug=settings.DEBUG)

    main(application)
        ''',
        'webservices_py':'''#########################################################
# Extending python path, file generated
#########################################################

def update_path():
    import os
    import sys

    base_dir = os.path.abspath(os.path.dirname(__file__))

    local_dir = os.path.join(base_dir,'%(local)s/')
    external_dir = os.path.join(base_dir,'%(external)s/')
    lib_dir = os.path.join(base_dir,'%(lib)s/')

    if lib_dir not in sys.path:
        import site
        site.addsitedir(lib_dir)

    if local_dir not in sys.path:
        sys.path.insert(0, local_dir)

    if external_dir not in sys.path:
        sys.path.insert(0, external_dir)

update_path()

from EchoServer_server import * # this class is generated by 'python wsdl2py SimpleEcho.wsdl' see http://carlitos-kyo.blogspot.com/2009/12/web-services-python-google-app-engine.html

# Implementing services

class wsEchoServer(EchoServer):
    # Make WSDL available for HTTP GET
    wsdl = "".join(open('SimpleEcho.wsdl').readlines())
    disco = "".join(open('SimpleEcho.disco').readlines())

    def soap_Echo(self, ps, **kw):
        request, response = EchoServer.soap_Echo(self, ps, **kw)
        return request, request

def main():
    application = WSGIApplication()
    application['EchoServer.asmx'] = SOAPAplication(wsEchoServer())
    #application['OtherServer.asmx'] = SOAPAplication(wsOtherServer()) # could have many webservices
    run_wsgi_app(application)

if __name__ == '__main__':

    # first setting which file use
    import os
    os.environ['turboengine.settings'] = 'project.%(settings)s'

    from turboengine.webservices.application import SOAPAplication
    from turboengine.webservices.application import WSGIAplication

    main()
        '''
    },

    'gae':{
        'app_py': '''#########################################################
# Extending python path, file generated
#########################################################

def update_path():
    import os
    import sys

    base_dir = os.path.abspath(os.path.dirname(__file__))

    local_dir = os.path.join(base_dir,'%(local)s/')
    external_dir = os.path.join(base_dir,'%(external)s/')
    lib_dir = os.path.join(base_dir,'%(lib)s/')

    if lib_dir not in sys.path:
        sys.path.insert(0, lib_dir)

    if local_dir not in sys.path:
        sys.path.insert(0, local_dir)

    if external_dir not in sys.path:
        sys.path.insert(0, external_dir)

update_path()

from google.appengine.ext import webapp
from google.appengine.ext.webapp.util import run_wsgi_app

class MainPage(webapp.RequestHandler):
    def get(self):
        self.response.headers['Content-Type'] = 'text/plain'
        self.response.out.write('Hello, webapp World!')

application = webapp.WSGIApplication(
                                     [('/', MainPage)],
                                     debug=True)

def main():
    run_wsgi_app(application)

if __name__ == "__main__":
    main()
        ''',
    },

    'common':{
        'app_yaml': '''application: YOUR_APP_NAME
version: 0
runtime: python
api_version: 1

handlers:
# Statics files
- url: /favicon.ico
  static_files: static/favicon.ico
  upload: static/favicon.ico

- url: /static
  static_dir: static

- url: /.*
  script: app.py

# decomment this if you will use webservices
#- url: /webservices/.*
#  script: app.py

skip_files:
- ^(.*/)?app\.yaml
- ^(.*/)?app\.yml
- ^(.*/)?index\.yaml
- ^(.*/)?index\.yml
- ^(.*/)?#.*#
- ^(.*/)?.*~
- ^(.*/)?.*\.py[co]
- ^(.*/)?.*/RCS/.*
- ^(.*/)?\..*
- ^(.*/)?.old\.*
- ^(.*/)?.docs\.*
        ''',
        'index_yaml': '''#indexes:

#- kind: Cat
#  ancestor: no
#  properties:
#  - name: name
#  - name: age
#    direction: desc
        ''',
        'cron_yaml': '''#cron:
#- description: daily summary job
#  url: /tasks/summary
#  schedule: every 24 hours
#- description: monday morning mailout
#  url: /mail/weekly
#  schedule: every monday 09:00
#  timezone: Australia/NSW
        ''',
        'queue_yaml': '''#queue:
#- name: default
#  rate: 1/s
#- name: mail-queue
#  rate: 2000/d
#  bucket_size: 10
#- name: background-processing
#  rate: 5/s
        ''',
        'dos_yaml': '''#blacklist:
#- subnet: 1.2.3.4
#  description: a single IP address
#- subnet: 1.2.3.4/24
#  description: an IPv4 subnet
#- subnet: abcd::123:4567
#  description: an IPv6 address
#- subnet: abcd::123:4567/48
#  description: an IPv6 subnet
        ''',
    }
}

base_html = '''{% load i18n %}<!DOCTYPE html>

<!--[if IE 7 ]><html lang="{{ LANGUAGE_CODE }}" class="no-js ie7"> <![endif]-->
<!--[if IE 8 ]><html lang="{{ LANGUAGE_CODE }}" class="no-js ie8"> <![endif]-->
<!--[if IE 9 ]><html lang="{{ LANGUAGE_CODE }}" class="no-js ie9"> <![endif]-->
<!--[if (gt IE 9)|!(IE)]><!--> <html lang="{{ LANGUAGE_CODE }}" class="no-js"> <!--<![endif]-->

<head>

  <meta charset="utf-8">
  
  {% if meta_description %}<meta name="description" content="{{ meta_description }}">{% endif %}
  {% if meta_keywords %}<meta name="keywords" content="{{ meta_keywords }}">{% endif %}
  <meta name="author" content="linkux-it">
  <meta name="robots" content="index, follow">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <!-- favicon 16x16 -->
  <link rel="shortcut icon" href="/{{ STATIC_URL }}images/favicon.ico">
  <!-- apple touch icon 57x57 -->
  <link rel="apple-touch-icon" href="{{ STATIC_URL }}images/favicon.png">
  
  <script type="text/javascript" src="http://code.jquery.com/jquery.min.js"></script> 
  
  <!-- Main style sheet. Change version number in query string to force styles refresh -->
  <!-- Link element no longer needs type attribute -->
  <link rel="stylesheet" href="css/screen.css?v=1.0">

  <!-- Modernizr for feature detection of CSS3 and HTML5; must be placed in the "head" -->
  <!-- Script tag no longer needs type attribute -->
  <script src="js/modernizr-1.6.min.js"></script>

  <!-- Remove the script reference below if you're using Modernizr -->
  <!--[if lt IE 9]>
  <script src="http://html5shiv.googlecode.com/svn/trunk/html5.js"></script>
  <![endif]-->
  
  {% block extra_base_head %}{% endblock %}
  
  {% block extra_head %}{% endblock %}

  <title>HTML5 template{% block title %}{% if title %} | {{ title }}{% endif %}{% endblock %}</title>
  
</head>

<body>
  
  <header>
  </header>

  <footer>
  </footer>
  
</body>

</html>
'''

t_404_html = '''{% extends 'base.html' %}
{% load i18n %}
{% block container %}
{% trans 'Page Not Found' %}
{% endblock %}'''

t_500_html = '''{% extends 'base.html' %}
{% load i18n %}
{% block container %}
{% trans 'Server error' %}
{% endblock %}'''


src/gaebuild/installer.py
import os
import re
import shutil
import urllib2
import templates
import setuptools
import subprocess
import zc.recipe.egg

from zc.buildout import UserError
from dist_installer import install

class Installer:

    def __init__(self, options, buildout, log, name):
        self.options = options
        self.buildout = buildout
        self.log = log
        self.name = name

    def get_extra_paths(self):
        basic_dir = os.path.join(self.buildout['buildout']['directory'], 'src')
        local_apps_dir = os.path.join(basic_dir, self.options['project'], self.options['local-apps'])
        external_apps_dir = os.path.join(basic_dir, self.options['project'], self.options['external-apps'])
        extra_paths = [self.options['location'],
                       self.buildout['buildout']['directory'],
                       local_apps_dir, external_apps_dir]

        pythonpath = [p.replace('/', os.path.sep) for p in
                      self.options['libs-path'].splitlines() if p.strip()]

        extra_paths.extend(pythonpath)

        return extra_paths

    def command(self, cmd, **kwargs):
        output = subprocess.PIPE
        if self.buildout['buildout'].get('verbosity'):
            output = None
        command = subprocess.Popen(
            cmd, shell=True, stdout=output, **kwargs)
        return command.wait()

    def create_file(self, file, template, options=None):
        f = open(file, 'w')
        if options is not None:
            f.write(template % options)
        else:
            f.write(template)
        f.close()

    def make_scripts(self, location, extra_paths, ws):
        scripts = []
        if self.options['turboengine'] != '':
            self.create_file(os.path.join(location, self.options['project'],'app.py'), templates.configs['turboengine']['app_py'], {'local': self.options['local-apps'],
                                                                                                                                    'external': self.options['external-apps'],
                                                                                                                                    'lib': self.options['libs-path'],
                                                                                                                                    'settings': self.options['settings']})
        return scripts

    def create_project(self, project_dir, project):
        old_config = self.buildout._read_installed_part_options()[0]
        if self.name in old_config:
            old_config = old_config[self.name]

            if 'project' in old_config and\
               old_config['project'] != self.options['project']:
                self.log.warning("GAEBuild: creating new project '%s', to replace previous project '%s'"%(self.options['project'], old_config['project']))

        # saving current work directory
        old_cwd = os.getcwd()
        os.chdir(project_dir)

        # importing current django instalation
        os.makedirs(project)
        os.chdir(project)

        os.makedirs('docs')

        os.makedirs('static/css')
        os.makedirs('static/js')
        os.makedirs('static/images')

        os.makedirs('templates')
        os.makedirs('i18n')

        self.create_file("templates/base.html", templates.base_html)
        self.create_file("templates/404.html", templates.t_404_html)
        self.create_file("templates/500.html", templates.t_500_html)

        self.create_file("app.yaml", templates.configs['common']['app_yaml'], {})
        self.create_file("index.yaml", templates.configs['common']['index_yaml'], {})
        self.create_file("cron.yaml", templates.configs['common']['cron_yaml'], {})
        self.create_file("queue.yaml", templates.configs['common']['queue_yaml'], {})
        self.create_file("dos.yaml", templates.configs['common']['dos_yaml'], {})

        if self.options['turboengine'] == '':
            self.create_file("app.py", templates.configs['gae']['app_py'], {'local': self.options['local-apps'],
                                                                            'external': self.options['external-apps'],
                                                                            'lib':self.options['libs-path']})

        if self.options['turboengine'] != '':
            os.makedirs('project')
            self.create_file('project/__init__.py', '', {})

            self.create_file("settings.py", '# Customize here settings', {})
            self.create_file("project/development.py", templates.development_settings, {})
            self.create_file("project/production.py", templates.production_settings, {})

            if self.options['webservices'].lower() == 'true':
                self.create_file("webservices.py", templates.configs['turboengine']['webservices_py'], {'local': self.options['local-apps'],
                                                                                                        'external': self.options['external-apps'],
                                                                                                        'lib':self.options['libs-path']})

        # updating to original cwd
        os.chdir(old_cwd)

    def update_project_structure(self):
        old_config = self.buildout._read_installed_part_options()[0]
        # updating old config to project name
        if self.name in old_config:
            old_config = old_config[self.name]

            if 'local-apps' in old_config and\
               old_config['local-apps'] != self.options['local-apps']:
                if os.path.exists(old_config['local-apps']):
                    self.log.info("GAEBuild: moving local-apps dir from % to %s"%(old_config['local-apps'], self.options['local-apps']))
                    shutil.move(old_config['local-apps'], self.options['local-apps'])

            if 'external-apps' in old_config and\
               old_config['external-apps'] != self.options['external-apps']:
                if os.path.exists(old_config['external-apps']):
                    self.log.info("GAEBuild: moving external-apps dir from % to %s"%(old_config['external-apps'], self.options['external-apps']))
                    shutil.move(old_config['external-apps'], self.options['external-apps'])

            if 'libs-path' in old_config and\
               old_config['libs-path'] != self.options['libs-path']:
                if os.path.exists(old_config['libs-path']):
                    self.log.info("GAEBuild: moving libs-path dir from % to %s"%(old_config['libs-path'], self.options['libs-path']))
                    shutil.move(old_config['libs-path'], self.options['libs-path'])

            if 'script-dir' in old_config and\
               old_config['script-dir'] != self.options['script-dir']:
                if os.path.exists(old_config['script-dir']):
                    self.log.info("GAEBuild: moving script-dir dir from % to %s"%(old_config['script-dir'], self.options['script-dir']))
                    shutil.move(old_config['script-dir'], self.options['script-dir'])

        if not os.path.exists(self.options['local-apps']):
            self.log.info("GAEBuild: creating local-apps dir %s"%(self.options['local-apps']))
            os.makedirs(self.options['local-apps'])

        if not os.path.exists(self.options['external-apps']):
            self.log.info("GAEBuild: creating external-apps dir %s"%(self.options['external-apps']))
            os.makedirs(self.options['external-apps'])

        if not os.path.exists(self.options['libs-path']):
            self.log.info("GAEBuild: creating libs-path dir %s"%(self.options['libs-path']))
            os.makedirs(self.options['libs-path'])

        answer = raw_input("Do you want to install/update apps?(yes/no): ")

        if answer.lower() == 'yes':
            print '\n************** Intalling gae/django apps **************\n'
            apps = self.options.get('apps', '').split()
            if len(apps) == 0:
                self.log.info('No apps to install')
            else:
                install_dir = os.path.abspath(self.options['external-apps'])

                args = ['-U', '-b', self.buildout['buildout']['download-cache'], '-d', install_dir]
                args.extend(apps)

                links = self.options.get('find-links', '').split()

                if len(links)>0:
                    links.insert(0, '-f')
                    args.extend(links)

                install(args)
            print '\n************** End intalling gae/django apps **************\n'

            print '\n************** Intalling python projects **************\n'

            apps = self.options.get('libs', '').split()
            turboengine = self.options.get('turboengine', '')
            if turboengine != '':
                if turboengine.lower() == 'last' :
                    apps.append("turboengine")
                else:
                    apps.append("turboengine==%s"%self.options.get('turboengine'))
                if self.options.get('webservices').lower() == 'true':
                    apps.append("ZSI")
                    apps.append("zope.interface")

            if len(apps) == 0:
                self.log.info('No apps to install')
            else:
                from setuptools.command.easy_install import main
                install_dir = os.path.abspath(self.options['libs-path'])

                if self.options.get('zipped').lower() == 'true':
                    args = ['-U', '-z', '-d', install_dir]
                else:
                    args = ['-U', '-d', install_dir]

                args.extend(['-s', self.options['script-dir']])
                links = self.options.get('find-links', '').split()

                if len(links)>0:
                    links.insert(0, '-f')
                    args.extend(links)

                args.extend(apps)
                previous_path = os.environ.get('PYTHONPATH', '')
                if previous_path == '':
                    os.environ['PYTHONPATH'] = '%s'%(install_dir)
                else:
                    os.environ['PYTHONPATH'] = '%s:%s'%(install_dir, previous_path)

                main(args)# installing libs

                if previous_path == '':
                    del os.environ['PYTHONPATH']
                else:
                    os.environ['PYTHONPATH'] = previous_path

            print '\n************** End intalling python projects **************\n'

    def verify_or_create_download_dir(self, download_dir):
        if not os.path.exists(download_dir):
            os.mkdir(download_dir)

    def install_recipe(self, location):
        self.options['setup'] = location
        development = zc.recipe.egg.Develop(self.buildout,
                                            self.options['recipe'],
                                            self.options)

        #development.install()
        del self.options['setup']

    def install_project(self, project_dir, project):
        if not os.path.exists(os.path.join(project_dir, project)):
            self.create_project(project_dir, project)
        else:
            self.log.info(
                'Skipping creating of project: %(project)s since '
                'it exists' % self.options)

        # creating structure
        old_cwd = os.getcwd()
        os.chdir(os.path.join(project_dir, project))

        self.update_project_structure()

        os.chdir(old_cwd)

    def install_scripts(self, location, extra_path, ws):
        script_paths = []

        # Make the wsgi and fastcgi scripts if enabled
        script_paths.extend(self.make_scripts(location, extra_path, ws))

        return script_paths
